harness/determined/pytorch/_pytorch_context.py

```diff
@@ -492,13 +543,31 @@ def clip_grads(params):
                 )
 
             if clip_grads is not None:
+                if self._scaler and self._auto_amp:
+                    self._scaler.unscale()
                 clip_grads(parameters)
 
-            if self.hvd_config.use:
-                with optimizer.skip_synchronize():  # type: ignore
-                    optimizer.step()
+            # For stepping the optimizer we will operate on the scaler passed
+            # in, or fall back to the wrapped scaler (if any).
+            if scaler is None:
+                scaler = self._scaler
+            if scaler:
+                if self.hvd_config.use:
+                    # TODO can we do a type assertion for horovod.torch._DistributedOptimizer
+                    with optimizer.skip_synchronize():  # type: ignore
+                        scaler.step(optimizer)
+                else:
+                    scaler.step(optimizer)
             else:
-                optimizer.step()
+                if self.hvd_config.use:
+                    # TODO can we do a type assertion for horovod.torch._DistributedOptimizer
+                    with optimizer.skip_synchronize():  # type: ignore
+                        optimizer.step()
+                else:
+                    optimizer.step()
+
+            if self._scaler and self._auto_amp:
+                self._scaler.update()
```

Comment:
Question: what happens if a user has only one scaler but many optimizers and losses?

I think we are doing it wrong to update the scaler every optimizer step often.  See https://pytorch.org/docs/master/notes/amp_examples.html#working-with-multiple-models-losses-and-optimizers

Reply:
Since we won't really know which one is the last one, IMO if you have multiple optimizers or losses then you need to use the manual API. I can add a comment to make that explicit. Thanks for the link, though - I had looked for something along those lines when I started this and hadn't seen that spelled out.

Reply:
And actually I can make sure we throw an error if you try to use the automatic API and multiple optimizers get wrapped.

Reply:
Wouldn't we just call update on the scaler once after train_batch finished to support that case?

Reply:
... yeah that would be the smart way to do it, huh. Done.